### PR TITLE
Graduate to yarn 4 🧶

### LINKED ISSUE #2228: Update yarn version
https://yarnpkg.com/blog/release/4.0#hardened-mode

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

- Updates yarn version to 4.0

### Related issues

- Fixes https://github.com/hyperlane-xyz/issues/issues/687
- Fixes #2228 
- Fixes https://github.com/hyperlane-xyz/issues/issues/704

### CODE REVIEW

**Review by aroralanuk — APPROVED** [review]

**Comment by aroralanuk** [comment]
https://github.com/hyperlane-xyz/hyperlane-monorepo/actions/runs/6803178647/job/18497901300?pr=2899

File: .lintstagedrc [hunk 1]
```diff
@@ -1,6 +1,5 @@
 {
   "*.js": "prettier --write",
   "*.ts": "prettier --write", 
-  "*.md": "prettier --write",
-  "*.sol": "prettier --write"
```
Comment (jmrossy) [inline .lintstagedrc:5]: I had to remove sol here b.c. the sol specific prettier settings are only setup in the core package, not across the monorepo. We can probably configure it higher up if we want. Tbh I'm not sure how this if/how this ever worked

File: typescript/infra/package.json [hunk 2]
```diff
@@ -67,5 +68,8 @@
     "node": "hardhat node",
     "prettier": "prettier --write *.ts ./src ./config ./scripts ./test",
     "test": "mocha --config ../sdk/.mocharc.json test/agents.test.ts"
+  },
+  "peerDependencies": {
```
Comment (jmrossy) [inline typescript/infra/package.json:72]: @yorhodes why did you add this? I don't think it will have any affect given that infra is a private (not published) package

File: typescript/sdk/package.json [hunk 3]
```diff
@@ -61,5 +63,9 @@
     "test:hardhat": "hardhat test $(find ./src -name \"*.hardhat-test.ts\")"
   },
   "types": "dist/index.d.ts",
-  "stableVersion": "1.5.0"
+  "stableVersion": "1.5.0",
+  "peerDependencies": {
+    "@ethersproject/abi": "*",
+    "@ethersproject/providers": "*"
+  }
```
Comment (jmrossy) [inline typescript/sdk/package.json:70]: Ditto

**Review by jmrossy — APPROVED** [review]

File: typescript/infra/package.json (hunk 2 above)
Comment (yorhodes) [reply typescript/infra/package.json:72]: it resolved some warnings yarn was giving
